File: wear/src/main/java/com/pdyjak/powerampwear/settings/SettingsManager.java
package com.pdyjak.powerampwear.settings;

import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import java.util.HashSet;
import java.util.Set;

public class SettingsManager {
    private static final String CIRCULAR_SCROLLING_KEY = "use_circular_scrolling";
    private static final String SHOW_CLOCK_KEY = "show_clock";
    private static final String VOLUME_CONTROLS_ONBOARDING_KEY = "vc_onboarding";
    private static final String QUICK_NAVIGATION_HINT_KEY = "quick_nav";

    public static class Listener {
        public void onCircularScrollingChanged() {}
        public void onClockSettingChanged() {}
    }

    @NonNull
    private final Set<Listener> mSettingsListeners = new HashSet<>();
    @NonNull
    private final SharedPreferences mPrefs;

    public SettingsManager(@NonNull SharedPreferences sharedPreferences) {
        mPrefs = sharedPreferences;
    }

    public boolean quickNavigationHintShown() {
        return mPrefs.getBoolean(QUICK_NAVIGATION_HINT_KEY, false);
    }

    public void saveQuickNavigationHintShown() {
        mPrefs.edit().putBoolean(QUICK_NAVIGATION_HINT_KEY, true).apply();
    }

    public boolean volumeControlsOnboardingShown() {
        return mPrefs.getBoolean(VOLUME_CONTROLS_ONBOARDING_KEY, false);
    }

    public void saveVolumeControlsOnboardingShown() {
        mPrefs.edit().putBoolean(VOLUME_CONTROLS_ONBOARDING_KEY, true).apply();
    }

    public boolean useCircularScrollingGesture() {
        return mPrefs.getBoolean(CIRCULAR_SCROLLING_KEY, false);
    }

    public boolean shouldShowClock() {
        return mPrefs.getBoolean(SHOW_CLOCK_KEY, true);
    }

    public void saveUseCircularScrolling(boolean enabled) {
        if (enabled == useCircularScrollingGesture()) return;
        mPrefs.edit().putBoolean(CIRCULAR_SCROLLING_KEY, enabled).apply();
        Set<Listener> copy = new HashSet<>(mSettingsListeners);
        for (Listener listener : copy) {
            listener.onCircularScrollingChanged();
        }
    }

    public void saveShowClock(boolean enabled) {
        if (enabled == shouldShowClock()) return;
        mPrefs.edit().putBoolean(SHOW_CLOCK_KEY, enabled).apply();
        Set<Listener> copy = new HashSet<>(mSettingsListeners);
        for (Listener listener : copy) {
            listener.onClockSettingChanged();
        }
    }

    public void addSettingsListener(@NonNull Listener listener) {
        mSettingsListeners.add(listener);
    }

    public void removeListener(@NonNull Listener listener) {
        mSettingsListeners.remove(listener);
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/settings/SettingsManager.kt
package com.pdyjak.powerampwear.settings

import android.content.SharedPreferences
import com.pdyjak.powerampwear.common.SimpleEvent

class SettingsManager(private val mPrefs: SharedPreferences) {

    companion object {
        private val CIRCULAR_SCROLLING_KEY = "use_circular_scrolling"
        private val SHOW_CLOCK_KEY = "show_clock"
        private val VOLUME_CONTROLS_ONBOARDING_KEY = "vc_onboarding"
        private val QUICK_NAVIGATION_HINT_KEY = "quick_nav"
    }

    val onCircularScrollingChanged = SimpleEvent()
    val onClockSettingChanged = SimpleEvent()

    fun quickNavigationHintShown(): Boolean {
        return mPrefs.getBoolean(QUICK_NAVIGATION_HINT_KEY, false)
    }

    fun saveQuickNavigationHintShown() {
        mPrefs.edit().putBoolean(QUICK_NAVIGATION_HINT_KEY, true).apply()
    }

    fun volumeControlsOnboardingShown(): Boolean {
        return mPrefs.getBoolean(VOLUME_CONTROLS_ONBOARDING_KEY, false)
    }

    fun saveVolumeControlsOnboardingShown() {
        mPrefs.edit().putBoolean(VOLUME_CONTROLS_ONBOARDING_KEY, true).apply()
    }

    fun useCircularScrollingGesture(): Boolean {
        return mPrefs.getBoolean(CIRCULAR_SCROLLING_KEY, false)
    }

    fun shouldShowClock(): Boolean {
        return mPrefs.getBoolean(SHOW_CLOCK_KEY, true)
    }

    fun saveUseCircularScrolling(enabled: Boolean) {
        if (enabled == useCircularScrollingGesture()) return
        mPrefs.edit().putBoolean(CIRCULAR_SCROLLING_KEY, enabled).apply()
        onCircularScrollingChanged()
    }

    fun saveShowClock(enabled: Boolean) {
        if (enabled == shouldShowClock()) return
        mPrefs.edit().putBoolean(SHOW_CLOCK_KEY, enabled).apply()
        onClockSettingChanged()
    }
}
